src/content/app/tools/vep/state/vep-action-listeners/vepSubmissionStatusPolling.ts

```diff
@@ -0,0 +1,182 @@
+  submissions: Record<string, PolledSubmission> = {};
+  queue: string[] = [];
+  isProcessing: boolean = false;
+  dispatch: AppDispatch | null = null;
+
+  enqueueSubmission = ({
+    submission,
+    dispatch
+  }: {
+    submission: PolledSubmission;
+    dispatch: AppDispatch;
+  }) => {
+    this.submissions[submission.id] = submission;
+    this.queue.unshift(submission.id);
+    this.dispatch = dispatch;
+
+    if (!this.isProcessing) {
+      this.processQueueWithDelay();
+    }
+  };
+
+  processQueueWithDelay = async () => {
+    if (!this.queue.length) {
+      this.isProcessing = false;
+      return;
+    }
+    this.isProcessing = true;
+
+    await this.pause();
+
+    const submissionId = this.queue.pop() as string;
+    const url = `${config.toolsApiBaseUrl}/vep/submissions/${submissionId}/status`;
+    const response = await fetch(url);
+    await this.handleResponse({ submissionId, response });
+
+    // this makes it a recursive function; but by avoiding the return statement, we shouldn't be leaking memory
+    this.processQueueWithDelay();
+  };
+
+  // method to run if there are any submissions with unresolved statuses in the browser storage
+  // when the app starts up
+  processSubmissionsOnStartup = async ({
+    submissions,
+    dispatch
+  }: {
+    submissions: PolledSubmission[];
+    dispatch: AppDispatch;
+  }) => {
+    this.dispatch = dispatch;
+
+    // query submission statuses one right after the other without the artificial delay,
+    // because it is quite likely that they have finished running
+    for (const submission of submissions) {
+      this.submissions[submission.id] = submission;
+      const submissionId = submission.id;
+      const url = `${config.toolsApiBaseUrl}/vep/submissions/${submissionId}/status`;
+      const response = await fetch(url);
+      await this.handleResponse({ submissionId, response });
+    }
+
+    // in case there are any submissions with unresolved statuses, continue polling their statuses as usual
+    this.processQueueWithDelay();
+  };
+
+  handleResponse = async ({
+    submissionId,
+    response
+  }: {
+    submissionId: string;
+    response: Response;
+  }) => {
+    if (!response.ok) {
+      if (response.status >= 500) {
+        // try later
+        this.queue.unshift(submissionId);
+      } else if (response.status === 404)
+        // fail submission
+        this.reportSubmissionStatus({
+          submissionId,
+          status: 'FAILED'
+        });
+    }
+
+    const { status } = await response.json();
+
+    if (!this.submissions[submissionId]) {
+      // this should not happen
+      return;
+    }
+
+    if (status !== this.submissions[submissionId].status) {
+      this.submissions[submissionId].status = status;
+      this.reportSubmissionStatus({
+        submissionId,
+        status
+      });
+    } else {
+      // put submission id back in the queue to query its status again later
+      this.queue.unshift(submissionId);
```

Comment:
Sorry I am trying get my head around this. Does this assume the queue will be FIFO?

Reply:
Yes. Items enter on the left, and move towards the right.